boards: infineon: Add support for Infineon PSOC4100T Plus based boards

manoj-aerlync opened the pull request:
Introduce CY8CPROTO_041TP EVK
Add support for Infineon CAT2 HAL

nordicjm reviewed (changes requested):
split into proper commits e.g. for adding soc, for adding board, look at https://docs.zephyrproject.org/latest/contribute/style/kconfig.html and read and apply rules before re-requesting review

sreeramIfx commented on modules/hal_infineon/mtb-pdl-cat2/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+set(pdl_dev_cat2_dir  ${pdl_dir}/devices)
+
+# Generate PDL specific SOC defines
+zephyr_compile_definitions(${CONFIG_SOC_PART_NUMBER})
```
Please use zephyr_library_compile_definitions(), More details @ https://github.com/zephyrproject-rtos/zephyr/pull/95874#discussion_r2413323138

dharunkrithikk replied:
changed

sreeramIfx commented on soc/infineon/cat2/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+add_subdirectory(${SOC_SERIES})
+zephyr_compile_definitions($<UPPER_CASE:${CONFIG_SOC}>_${CONFIG_SOC_PART_NUMBER})
```
Is this required? is it not taken care by the infineon_kconfig.h?

Deepika-aerlync replied:
This line is removed from the CMakeLists.txt.

JarmouniA reviewed (commented):
The manifest update commit should be 1st.

sreeramIfx commented on dts/bindings/arm/infineon,cat2-scb.yaml:
```diff
@@ -0,0 +1,20 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: Infineon Serial Communication Blocks (SCB) node
```
Why do we need a separate binding file for UART?

jsbatch commented on boards/infineon/cy8cproto_041tp/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+.. target-notes::
+
+.. _cy8cproto_041tp Board Website:
+   https://www.infineon.com/assets/row/public/documents/30/57/infineon-psoc-4100t-plus-architecture-trm-additionaltechnicalinformation-en.pdf
```
This link points to the TRM for the PSOC 4100T Plus MCU, not the board website.

Deepika-aerlync replied:
Placed the link for cy8cproto_041tp Board Website

jsbatch commented on drivers/clock_control/Kconfig.ifx_cat2:
This file looks like a duplicate of options defined in Kconfig.ifx_cat1, so I don't think this file is necessary.

jsbatch commented on drivers/serial/uart_ifx_cat1_pdl.c:
```diff
@@ -216,32 +145,28 @@ cy_rslt_t ifx_cat1_uart_set_baud(const struct device *dev, uint32_t baudrate)
 
-	best_oversample = best_oversample;
-
 	data->scb_config.oversample = best_oversample;
```
Should this be based on an actual calculation similar to best_difference?  If not, just use the defined value directly.

sayoojkkarun replied:
This is an already merged code. Can we have this as another pull request if this needs changes?

jsbatch commented on boards/infineon/cy8cproto_041tp/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+    https://softwaretools.infineon.com/tools/com.ifx.tb.tool.modustoolbox
+
+.. _ModusToolbox Programming Tools:
+    https://softwaretools.infineon.com/tools/com.ifx.tb.tool.modustoolboxprogtools
```
https://www.infineon.com/design-resources/development-tools/sdk/modustoolbox-software/modustoolbox-programming-tools

jsbatch commented on drivers/serial/uart_ifx_cat1_pdl.c:
```diff
@@ -628,28 +569,32 @@ static const cy_stc_scb_uart_config_t _uart_default_config = {
 	.txFifoTriggerLevel = 63UL,
+#else
+	.txFifoTriggerLevel = 0,
+#endif
```
Rather than use hardcoded magic numbers here that differ between parts, I'd rather see #define values.  I think it makes the structure definition more readable.

Deepika-aerlync replied:
Replaced the hard-coded values with macros.

jsbatch commented on dts/arm/infineon/cat2/psoc4100tp/psoc4100tp.44-tqfp.dtsi:
Pinmux definitions are missing for 44-pin and 48-pin parts.

jsbatch commented on dts/arm/infineon/cat2/psoc4100tp/psoc4100tp.48-qfn.dtsi:
Pinmux definitions are missing for 44-pin and 48-pin parts.

jsbatch commented on include/zephyr/drivers/serial/uart_ifx.h:
```diff
@@ -0,0 +1,53 @@
+
+#define IFX_UART_MEM_WIDTH(cfg)                                                                    \
+	(((cfg).dataWidth <= CY_SCB_BYTE_WIDTH) ? (CY_SCB_CTRL_MEM_WIDTH_BYTE)                     \
+	 : (CY_SCB_CTRL_MEM_WIDTH_HALFWORD))
```
If these are being pulled out of the code, I think it would be better to turn them into inline functions instead of macros.  I find inline functions to be more readable and more maintainable.

Deepika-aerlync replied:
Updated the code to replace the macros with static inline functions as recommended.

Deepika-aerlync committed:
manifest: hal_infineon: update rev to latest

Update hal_infineon revision to add PSOC4 family support

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync committed:
modules: hal_infineon: add PSoC4 Kconfig support

Add Kconfig configuration support for PSoC4 family in hal_infineon module.

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync committed:
modules: hal_infineon: add PSoC4 build support

Add CMake build configuration support for PSoC4 PDL and templates.

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync committed:
dts: arm: infineon: add PSoC4100TP base devicetree

Add base devicetree support for Infineon PSoC4100TP series including:
- Core SoC devicetree (psoc4100tp.dtsi)
- CM0+ core configuration (psoc4100tp.cm0p.dtsi)
- Pin package variants (44-TQFP, 48-QFN, 48-TQFP, 64-TQFP)
- System clocks configuration (system_clocks.dtsi)

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync committed:
dts: arm: infineon: add CY8C4146 MPN variants

Add devicetree support for CY8C4146 MPN variants including:
- AXI packages (T403, T413, T453)
- AZI packages (T403-T455)
- AZQ packages (T413, T453)
- LQI packages (T403, T413, T453)

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync committed:
soc: infineon: add PSoC4100TP SOC support

Add SoC port for the Infineon PSoC4100TP series, including:
- Core initialization
- Kconfig files
- YAML device definitions
- SoC-specific pin control configuration

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync committed:
boards: infineon: add CY8CPROTO-041TP board support

Add board porting for the Infineon CY8CPROTO-041TP evaluation kit:
- Board devicetree and pinctrl configuration
- Kconfig and YAML definitions
- Documentation
- Support for west build, west flash, and west debug (OpenOCD)

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync committed:
drivers: clock_control: add Infineon PSOC4 support

Add clock control driver support for Infineon PSoC4:
- Fixed clock driver (clock_control_ifx_fixed_clock.c)
- Fixed factor clock driver (clock_control_ifx_fixed_factor_clock.c)
- Peripheral clock driver (clock_control_ifx_peri_clock.c)
- CAT1 clock control header

Provides clock initialization and management based on
devicetree configuration.

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync committed:
drivers: serial: add Infineon PSOC4 UART support

Add initial implementation of the Infineon PSOC4 UART driver based on the
Peripheral Driver Library (PDL) for PSOC4 (PSoC 4100TP) devices.

Features:
- UART transmit and receive using PDL SCB UART APIs
- Configurable baud rate, parity, stop bits, and data bits from devicetree
- Pinctrl integration for TX/RX pins
- Support for Zephyr console and shell subsystems
- Build-time validation for UART configuration parameters

Signed-off-by: Dharun krithik k <dharunkrithik@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync commented on boards/infineon/cy8cproto_041tp/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+.. _cy8cproto_041tp Board Website:
+   https://www.infineon.com/assets/row/public/documents/30/57/infineon-psoc-4100t-plus-architecture-trm-additionaltechnicalinformation-en.pdf
+
+.. _ModusToolbox:
```
Placed given link for ModusToolbox

Deepika-aerlync commented on include/zephyr/drivers/serial/uart_ifx.h:
```diff
@@ -0,0 +1,53 @@
+	(((sb) <= UART_CFG_STOP_BITS_2) ? stop_bits_lut[(sb)] : CY_SCB_UART_STOP_BITS_1)
+
+#define CONVERT_UART_DATA_BITS_Z_TO_CY(db)                                                         \
+	(((db) <= UART_CFG_DATA_BITS_9) ? data_bits_lut[(db)] : 1U)
```
Updated the code to use LOG_WRN

Deepika-aerlync commented on modules/hal_infineon/Kconfig:
```diff
@@ -115,8 +115,28 @@ config USE_INFINEON_SMIF
+config USE_INFINEON_SMARTIO
+	bool
+	help
+	Enable Smart-IO module driver for Infineon devices
```
Dropped above mentioned peripherals.

dharunkrithikk commented on soc/infineon/cat2/psoc4100tp/Kconfig.soc:
```diff
@@ -0,0 +1,314 @@
+	default "psoc4100tp" if SOC_SERIES_PSOC4100TP
+
+# SOC die
+config SOC_DIE_PSOC4100T_PLUS
```
changed

dharunkrithikk commented on west.yml:
```diff
@@ -185,7 +185,7 @@ manifest:
         - hal
     - name: hal_infineon
-      revision: f3c571f772209b5970bdd1806da641244b5c4c38
+      revision: 2d37de92236ea62b80562e7381ff43d9522dbbad
```
got merged so no changes done

dharunkrithikk commented on drivers/clock_control/clock_control_ifx_fixed_clock.c:
```diff
@@ -86,25 +86,39 @@ static int fixed_rate_clk_init(const struct device *dev)
 	case IFX_IHO:
 		Cy_SysClk_IhoEnable();
 		break;
+#endif
```
used dts to check, changed

Deepika-aerlync commented on soc/infineon/psoc4/psoc4100tp/Kconfig.soc:
```diff
@@ -0,0 +1,268 @@
+
+config SOC_CY8C4146AZI_T405
+	bool
+	select SOC_SERIES_PSOC4100TP
```
Added SOC_PACKAGE_X for missed devices.

Deepika-aerlync commented on drivers/serial/uart_ifx_cat1_pdl.c:
```diff
@@ -26,13 +29,14 @@
+#define IFX_UART_RX_FIFO_TRIGGER_LEVEL_A 63UL
+#define IFX_UART_TX_FIFO_TRIGGER_LEVEL_A 63UL
+#define IFX_UART_RX_FIFO_TRIGGER_LEVEL_B 7
+#define IFX_UART_TX_FIFO_TRIGGER_LEVEL_B 0
```
Removed

Deepika-aerlync commented on boards/infineon/cy8cproto_041tp/cy8cproto_041tp.yaml:
```diff
@@ -0,0 +1,17 @@
+  - zephyr
+  - gnuarmemb
+supported:
+  - uart
```
gpio is added.

Deepika-aerlync commented:
Okay @Sayooj Karun ***@***.***> will Update it

On Thu, Nov 27, 2025 at 2:21 PM Sayooj ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In boards/infineon/cy8cproto_041tp/cy8cproto_041tp.yaml
> <https://github.com/zephyrproject-rtos/zephyr/pull/98564#discussion_r2567650329>
> :
>
> > +# Copyright (c) 2025 Infineon Technologies AG,
> +# or an affiliate of Infineon Technologies AG.
> +#
> +# SPDX-License-Identifier: Apache-2.0
> +
> +identifier: cy8cproto_041tp
> +name: CY8CPROTO-041TP Development Board
> +type: mcu
> +arch: arm
> +ram: 32
> +flash: 128
> +toolchain:
> +  - zephyr
> +  - gnuarmemb
> +supported:
> +  - uart
>
> Hi @Deepika-aerlync <https://github.com/Deepika-aerlync> We cannot add
> gpio, as the driver is not present in this pull request.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/98564#discussion_r2567650329>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BYHFV7K7U32LZM7SIVTGFH33623Q3AVCNFSM6AAAAACKUY73NCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTKMJUGEYDGNRSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

manoj-aerlync committed:
dts: arm: infineon: add CY8C4147 MPN variants

Add devicetree support for CY8C4147 MPN variants including:
- AXI packages (T403-T493)
- AZI packages (T403-T475)
- AZQ packages (T415-T495)
- LQI packages (T403-T473)
- LQQ package (T493)

Signed-off-by: Manojkumar Konisetty <manoj@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>
Signed-off-by: Deepika aerlync <deepika@aerlync.com>

Deepika-aerlync commented on soc/infineon/psoc4/psoc4100tp/Kconfig.soc:
```diff
@@ -0,0 +1,279 @@
+	default "psoc4100tp" if SOC_SERIES_PSOC4100TP
+
+# SOC Packages for Infineon PSOC4100Tp series MCUs
+config SOC_PACKAGE_PSOC4100TP_TQFP_64
```
Changed.

mcatee-infineon reviewed (commented):
Looks good to me.

nordicjm commented on soc/infineon/psoc4/psoc4100tp/Kconfig.soc:
```diff
@@ -0,0 +1,279 @@
+	default "psoc4100tp" if SOC_SERIES_PSOC4100TP
+
+# SOC Packages for Infineon PSOC4100Tp series MCUs
+config SOC_PACKAGE_PSOC4100TP_64_TQFP
```
`SOC_PACKAGE_*` Kconfigs to be moved to `Kconfig` along with their selection (the SOC_SERIES_* selection remains in this file)

teburd commented on include/zephyr/drivers/serial/uart_ifx.h:
This should absolutely not be a public header file in Zephyr, this is part specific functionality being moved from the driver to a public header. Please move functions back into the driver .c file.

teburd commented on modules/hal_infineon/CMakeLists.txt:
```diff
@@ -98,3 +99,17 @@ endif()
   add_subdirectory(bless)
 endif()
+
+if(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
```
Add directly the needed cmake commands here, mtb named subdirectories with 4-5 line CMake files makes no sense in the context of Zephyr

Put the content of those inline here when selecting PSOC4, drop any reference to cat/mtb naming possible.

teburd commented on drivers/clock_control/clock_control_ifx_fixed_clock.c:
```diff
@@ -128,20 +128,33 @@ static int fixed_rate_clk_init(const struct device *dev)
 	case IFX_IMO:
-		break;
+		Cy_SysClk_ImoEnable();
+#if defined(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
```
This should really be a hidden Kconfig informing code that PSOC4 has an adjustable clock here.

teburd commented on boards/infineon/cy8cproto_041tp/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+Flashing
+========
+
+The KIT_XMC72_EVK includes an onboard programmer/debugger (`KitProg3`_) to provide debugging, flash programming, and serial communication over USB. Flash and debug commands use OpenOCD and require a custom Infineon OpenOCD version, that supports KitProg3, to be installed.
```
```suggestion
The onboard programmer/debugger (`KitProg3`_) provides debugging, flashing, and serial communication over USB. Flash and debug commands use OpenOCD and require a custom Infineon OpenOCD version, that supports KitProg3 to be installed.
```

teburd commented on dts/arm/infineon/psoc4/psoc4100tp/psoc4100tp.44-tqfp.dtsi:
```diff
@@ -0,0 +1,550 @@
+/ {
+	soc {
+		/delete-node/ gpio@40040500; // gpio_prt5
+		/delete-node/ gpio@40040600; // gpio_prt6
```
```suggestion
		/delete-node/ gpio@40040500; /* gpio_prt5 */
		/delete-node/ gpio@40040600; /* gpio_prt6 */
```

teburd commented on dts/arm/infineon/psoc4/psoc4100tp/psoc4100tp.48-qfn.dtsi:
```diff
@@ -0,0 +1,641 @@
+
+/ {
+	soc {
+		/delete-node/ gpio@40040600; // gpio_prt6
```
```suggestion
		/delete-node/ gpio@40040600; /* gpio_prt6 */
```

teburd commented on modules/hal_infineon/mtb-pdl-cat2/CMakeLists.txt:
```diff
@@ -0,0 +1,69 @@
+zephyr_library_sources(${pdl_drv_dir}/source/cy_sysint.c)
+zephyr_library_sources(${pdl_drv_dir}/source/cy_flash.c)
+zephyr_library_sources(${pdl_drv_dir}/source/cy_wdt.c)
+zephyr_library_sources(${pdl_drv_dir}/source/cy_tcpwm_pwm.c)
```
This should be only built when TCPWM is needed

teburd reviewed (changes requested):
The use of DT_CAT1_PINMUX() macros throughout the dts[i] files here is confusing given the PR title adding CAT2. If the macro is shared among various Infineon parts this should be renamed DT_IFX_PINMUX. This can happen in another PR but I do expect to see it.

manoj-aerlync commented on drivers/serial/uart_infineon_pdl.c:
```diff
@@ -675,18 +627,22 @@ static const cy_stc_scb_uart_config_t _uart_default_config = {
 	.breakWidth = 11UL,
 	.dropOnFrameError = false,
 	.dropOnParityError = false,
+#if !defined(CONFIG_SOC_SERIES_PSOC4100TP)
```
@sreeramIfx We need to make this change in hal/infineon. Shall we make this change?

nordicjm commented on modules/hal_infineon/mtb-template-cat2/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+set(psoc4_dir ${template_dir}/files/templates)
+
+if(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
+    zephyr_include_directories(${psoc4_dir}/COMPONENT_MTB)
```
2 space indent for cmake

nordicjm commented on modules/hal_infineon/infineon_kconfig.h:
```diff
@@ -44,4 +44,335 @@
+ */
+
+#if defined(CONFIG_SOC_CY8C4146LQI_T403)
+#ifndef CY8C4146LQI_T403
```
I have just followed the include path for this which is applications -> soc.h -> this file -> now has the equivalent of a lot of zephyr_library_definitions() applied which it should not have, this seems to be a file for your HAL, so soc.h should not include this. Why do applications or zephyr libraries (not HAL) need any of these?

I.e. you are introducing what this issue tries to remove: https://github.com/zephyrproject-rtos/zephyr/issues/98647

manoj-aerlync replied:
@sreeramIfx I think we need to make changes in HAL to fix this.

teburd reviewed (changes requested):
My comments have been addressed

teburd reviewed (dismissed):
My comments have really been addressed

nordicjm reviewed (commented):
final change needed

sayoojkkarun commented:
@sreeramIfx @teburd Could you please check the pull request again.
